### ISSUE TITLE
Wrapped optionals should probably collapse.

### ISSUE THREAD

**Issue by Andersama** [hanickadot/compile-time-regular-expressions#128]
In playing around with some regexs that have optionals (to work out maybe some fixes for the long compile times). I found that `(?:a?)?` generates `ctre::optional<ctre::optional<ctre::character<97>>>` when it should probably generate or evaluate as `ctre::optional<ctre::character<97>>`.

EG: probably should have
```c
// wrapped optional
template <typename R, typename Iterator, typename EndIterator, typename... Content, typename... Tail>
constexpr CTRE_FORCE_INLINE R evaluate(const Iterator begin, Iterator current, const EndIterator end, R captures, ctll::list<optional<optional<Content...>>, Tail...>) noexcept {
	if (auto r1 = evaluate(begin, current, end, captures, ctll::list<sequence<Content...>, Tail...>())) {
		return r1;
	} else if (auto r2 = evaluate(begin, current, end, captures, ctll::list<Tail...>())) {
		return r2;
	} else {
		return not_matched;
	}
}

// wrapped lazy optional
template <typename R, typename Iterator, typename EndIterator, typename... Content, typename... Tail> 
constexpr CTRE_FORCE_INLINE R evaluate(const Iterator begin, Iterator current, const EndIterator end, R captures, ctll::list<lazy_optional<lazy_optional<Content...>>, Tail...>) noexcept {
	if (auto r1 = evaluate(begin, current, end, captures, ctll::list<Tail...>())) {
		return r1;
	} else if (auto r2 = evaluate(begin, current, end, captures, ctll::list<sequence<Content...>, Tail...>())) {
		return r2;
	} else {
		return not_matched;
	}
}

// wrapped optional in lazy optional (optional processes first, supersedes lazy inside)
template <typename R, typename Iterator, typename EndIterator, typename... Content, typename... Tail>
constexpr CTRE_FORCE_INLINE R evaluate(const Iterator begin, Iterator current, const EndIterator end, R captures, ctll::list<lazy_optional<optional<Content...>>, Tail...>) noexcept {
	if (auto r1 = evaluate(begin, current, end, captures, ctll::list<sequence<Content...>, Tail...>())) {
		return r1;
	} else if (auto r2 = evaluate(begin, current, end, captures, ctll::list<Tail...>())) {
		return r2;
	} else {
		return not_matched;
	}
}

// wrapped lazy optional in optional (lazy optional processes first, supersedes optional inside)
template <typename R, typename Iterator, typename EndIterator, typename... Content, typename... Tail> 
constexpr CTRE_FORCE_INLINE R evaluate(const Iterator begin, Iterator current, const EndIterator end, R captures, ctll::list<optional<lazy_optional<Content...>>, Tail...>) noexcept {
	if (auto r1 = evaluate(begin, current, end, captures, ctll::list<Tail...>())) {
		return r1;
	} else if (auto r2 = evaluate(begin, current, end, captures, ctll::list<sequence<Content...>, Tail...>())) {
		return r2;
	} else {
		return not_matched;
	}
}
```

**Comment by hanickadot** [comment]
Fixed

**Comment by Andersama** [comment]
Does this handle lazy_optional<optional<Content...>> or optional<lazy_optional<Content...>>?

**Comment by hanickadot** [comment]
what would be semantic for that?

**Comment by hanickadot** [comment]
I propose to rewrite:
`(?:x?)??` -> `x??`
`(?:x??)?` -> `x?'

**Comment by Andersama** [comment]
I think that's flipped? I'm still lost on the ll1 parser. The inside part ends up superseding the outside so I think its.
`(?:x?)??` -> x?
`(?:x??)?` -> `x??`

**Comment by hanickadot** [comment]
I looked at PCRE and it behaves when either of those optional are lazy result is lazy.

**Comment by hanickadot** [comment]
Fixed in 73cee574a1efc4007c9ea1a27b61cf25da34983b

**Comment by Andersama** [comment]
Ah...nvm yeah I think you're right. the lazy quantifier would end up applying to them all.
